[DO NOT MERGE] Trying cascading3

cchepelov opened the pull request:
A quick-n-dirty attempt at using Cascading3 in scalding, at the source level. 

At this point, this will choke quickly based on a few API that either disappeared or moved within Cascading; specifically, FlowStep#getSources and FlowStep#getGraph. 

Really only a quick RFC at this point, will need to investigate how to best to support the "missing" APIs 
(suspecting this is really a matter of adapting to the new API made just to help building ExecutionContext.getDesc in a nicer way)

Commit:
TemplateTap had been deprecated for a while, dropped now. Drop TemplateSource.

(hoping that's right)

Commit:
a few .gitignores (test data)

Commit:
Begin attempting the jump to Cascading 3.0.2 !

Commit:
One property disappeared, got renamed. Follow suit + @deprecated

Commit:
Tap-n-Scheme generic definition slightly changed upstream, follow.

Commit:
CascadingStats is now CascadingStats[_]

Commit:
FlowStep generic definition changed as well

Commit:
(possibly reckless) attempt at telling Scala which macro interface we'd like to use

(really unsure of what doing here)

ianoc commented on scalding-core/src/main/scala/com/twitter/scalding/Config.scala:
```diff
@@ -101,15 +103,21 @@ trait Config extends Serializable {
     this + (SpillableProps.MAP_THRESHOLD -> count.toString)
 
+  @deprecated("deprecated in Cascading 2.7 and dropped in Cascading 3.0, use setMapSideAggregationCapacity", "cascading 2.7")
+  def setMapSideAggregationThreshold(count: Int): Config = {
```
This isn't really correct for scalding, the comment mentions the method is realted to the associated operations which wires down to https://github.com/twitter/scalding/blob/develop/scalding-core/src/main/scala/com/twitter/scalding/Operations.scala#L126 which uses the same key as cascading. Not sure we shouldn't just be keeping the method/key as it was just defining the constant in scalding instead

johnynek replied:
+1 to what @ianoc said. I'd rather us take the max value of the two keys in Operations.scala, and here set the non-deprecated internally, but not deprecate our API (instead explain what it means to the user, not necessarily in cascading terms).

johnynek replied:
Why not call the capacity variant?

ianoc commented on scalding-core/src/main/scala/com/twitter/scalding/Operations.scala:
```diff
@@ -134,7 +134,7 @@ package com.twitter.scalding {
 
     val DEFAULT_CACHE_SIZE = 100000
-    val SIZE_CONFIG_KEY = AggregateBy.AGGREGATE_BY_THRESHOLD
+    val SIZE_CONFIG_KEY = AggregateByProps.AGGREGATE_BY_CAPACITY
```
This is going to break lots of users, so lets try not do this. any usage of -D on the commandline will be broken

johnynek replied:
I'd rather take the max in cacheSize (maybe logging a warning if two are set).

cchepelov replied:
Trick is, AggregateBy.AGGREGATE_BY_THRESHOLD has disappeared from the Cascading API. We can of course read it and carry that value forward if AggregateByProps.AGGREGATE_BY_CAPACITY has been left unset by the user; is that what you'd like me to do?

Do you want me to bring that constant back (within scalding only), take AggregateByProps.AGGREGATE_BY_CAPACITY if defined or else take AggregateBy.AGGREGATE_BY_THRESHOLD (logging a warning) ?

@johnynek just to make sure I understand, when you write " (instead explain what it means to the user, not necessarily in cascading terms)", do you mean logging a warning, or leaving a comment in the code?

ianoc commented on scalding-serialization/src/main/scala/com/twitter/scalding/serialization/macros/impl/OrderedBufferableProviderImpl.scala:
```diff
@@ -16,7 +16,7 @@
 
 import scala.language.experimental.macros
-import scala.reflect.macros.Context
+import scala.reflect.macros.whitebox.Context
```
Please undo non scala 2.10 compatible changes, they don't seem like they should be in this PR anyway

johnynek replied:
we can possibly use https://github.com/milessabin/macro-compat if we want to escape any 2.11/2.12 warnings here, but I'd really like to keep 2.10 support as long as we can easily do so (even though twitter is actually on 2.11 these days).

cchepelov replied:
Indeed, this was reckless. Reverting now.

cchepelov commented:
Thanks for the prompt feedback, @ianoc and @johnynek ! Will address the various points now, then proceed with the rest

Commit:
Revert "(possibly reckless) attempt at telling Scala which macro interface we'd like to use"

This reverts commit 92f6abc1e97fd9b64270e0099507075ea91cd29b.

Commit:
Cascading 2→3 generic API changes

Signed-off-by: Cyrille Chépélov (TP12) <cch@transparencyrights.com>

Commit:
TemplateSource had to go, so should the test

Signed-off-by: Cyrille Chépélov (TP12) <cch@transparencyrights.com>

Commit:
Can't reference changed jgrapht internals anymore

Signed-off-by: Cyrille Chépélov (TP12) <cch@transparencyrights.com>

Commit:
steps no longer have numbers, they have an ID

Commit:
(not sure about 2.10) scala complains about "useless" @transient declaration

(probably not the best PR to do this, nor the best thing; >50% change I'll revert this anyway)

Commit:
Adjust to some Cascading API changes (getGraph→getElementGraph)

Commit:
actually use the setMapSideAggregationThreshold→Capacity terminology change

Commit:
reference cascading3-enabled serialization libraries (all pull requests for now...)

cchepelov commented:
… Right now, busy updating quite a few dependencies to Cascading3 (cascading-parquet, cascading-avro, elephant-bird-cascading3, etc. 
To be continued!

Commit:
(JobConf)→(_ <: Configuration)

Commit:
Update some expected error messages coming from Cascading 3.0

Commit:
Reinstate scala 2.10.5 cross-build

Commit:
Record the Elephant-bird snapshot's pull request

cchepelov commented:
@ianoc @johnynek Done for now. Trouble is, this requires that code from a few dependencies' pull requests be available at compile time, and I'm not quite sure how best to convince Travis to do this:

* https://github.com/ScaleUnlimited/cascading.avro/pull/44
* https://github.com/apache/parquet-mr/pull/284
* https://github.com/nathanmarz/dfs-datastores/pull/53
* https://github.com/twitter/elephant-bird/pull/454

For that reason, the code currently fails to run on Travis.

Commit:
Updating some upstream deps versions + exposing prebuilt dependencies' jars on conjars.org

Commit:
Following advice from cwensel, depending on altered groupId/artifactId where dependencies are patched too
(this is temporary)

Commit:
temporary groupId, yes, but correct temporary groupId

Commit:
(rollbacking the temp.cchepelov.* groupId namespace, causes too much busy work)
 + forgotten "cascading2" namespace in TypedJson

Commit:
Bumping source-compatibility to Java 8 was as bad an idea as ninja-dropping scalding 2.10. Oops.

Commit:
(wip) add travis_wait calls to dodge 10-minute timeouts

Commit:
(wut? travis_wait is a bash function it seems)

Commit:
Revert "(wut? travis_wait is a bash function it seems)"

This reverts commit 27f1d9aabbcb303bb75eaecbfde4485cf9eddd29.

Commit:
Revert "(wip) add travis_wait calls to dodge 10-minute timeouts"

This reverts commit 5bbdde7ef37c491c6e17bc4a3ac73c7f74e2f5bf.

Commit:
updating to latest released version of Cascading 3.0.x

cchepelov commented:
Now working off binaries dumped onto conjars.org for patched dependencies.

There is an outstanding problem at test stage: some Test Jobs seem to deadlock, apparently within the Cascading test driver. Investigating this week (hopefully).

cchepelov commented:
synced with 0.15.1-RC1+ at 55f03f7e2ba363b26556b5390c77220fdf8ee637

cchepelov commented:
The deadlock in SkewJoinPipeTest should hopefully be fixed by https://github.com/cwensel/cascading/pull/47 (unless that one breaks more things elsewhere ☺).

Commit:
new InvalidSourceTap also needs Tap interface changes

Commit:
new: ConfigBridge, in order to deal with the various [Config] types FlowSteps et al. can carry

(before Cascading 3.0, this was a JobConf on -hadoop, Properties on -local)

Commit:
cascading 3.0: Flow[JobConf] → Flow[_] where _ can be JobConf, Configuration, Property...

Commit:
Adjust how Pipe description texts are carried + update tests

Commit:
Adjust how the numReducers property is accessed (cascading 3.0)

Commit:
cascading 3.0 flow step id is now alphanumeric, not numeric

Commit:
update to currently locally-built cascading WIP. FIXME ASAP: use WIP post https://github.com/cwensel/cascading/pull/47

cchepelov commented:
Update: some tests fixed with respect to .WithDescription("hi there")
(9896ed9 will not build on Travis for now, awaiting for successful integration of https://github.com/cwensel/cascading/pull/47 )

cc @simondumas

Commit:
Cheat a bit on the dfs-datastores VersionedKeyValSourceTest

dfs-datastores' semantics have shifted since 1.3.4, and now it is of the consumer's responsibility to create
the empty version directory which will contain the version's data.
In this whitebox test, nobody was creating the empty directory anymore, so the test failed.

Added a "hint" and an explanation (not sure this tests much now, but keeps the original dance mostly in)

Commit:
Reductor estimate test failed, as Hadoop2.6 understands the legacy properties but responds in the current

johnynek commented on project/Build.scala:
```diff
@@ -45,15 +45,18 @@ object ScaldingBuild extends Build {
     organization := "com.twitter",
 
-    scalaVersion := "2.10.6",
+    scalaVersion := "2.10.5",
```
any reasons for this?

cchepelov replied:
no, oversight. reverting.

johnynek commented on project/Build.scala:
```diff
@@ -319,15 +324,16 @@ object ScaldingBuild extends Build {
-      "com.backtype" % "dfs-datastores-cascading" % dfsDatastoresVersion,
-      "com.backtype" % "dfs-datastores" % dfsDatastoresVersion,
+      "org.apache.hadoop" % "hadoop-client" % hadoopVersion % "provided",
+      "temp.cchepelov.com.backtype" % "dfs-datastores-cascading3" % dfsDatastoresVersion, // FIXME: https://github.com/nathanmarz/dfs-datastores/pull/53
```
#1473 should fix this

cchepelov replied:
#1473 sounds like a good idea, do you want me to pull https://github.com/rubanm/scalding/tree/rubanm/dfs_datastores into this?

rubanm replied:
@cchepelov  I'm running some tests against that branch, so it should get merged to develop soonish. Go ahead and merge that branch if that keeps this PR moving though. Thanks.

johnynek commented on scalding-core/src/main/scala/com/twitter/scalding/ConfigBridge.scala:
```diff
@@ -0,0 +1,51 @@
+   * This adapter handles the various vessel types that can be used to configure properties of a a
+   * FlowStep[_] since Cascading 3.0
+   */
+  class FlowStepAdapter(flowStep: FlowStep[_]) {
```
extends Any

cchepelov committed:
Adjusting dependencies (new locally-built cascading-3.1-wip-dev)
	+ cancel accidental downgrade of scala 2.10

cchepelov committed:
per @posco's suggestion, FlowStep[_ <: Any] not FlowStep[_]

cchepelov committed:
RFC: merge of grouped++streamed no longer supported by Cascading 3.0. Alternative tests to resolve this

cchepelov committed:
Using cascading-3.1-wip-47

Commit:
Adjusting expected test line numbers (yearning for C++'s good old __LINE__ macro...)

Commit:
bump Hadoop to 2.7.1, pulled in by Cascading anyway

Commit:
(WIP: opening up hadoop2-mr1 and tez tests)

cchepelov committed:
(WIP; fixing some compile errors)

Commit:
Disable tests which will not work in Tez

Commit:
Revisit fabric-switching logic to also work in Mode#openForRead (required by tests)

Commit:
The Tez fabric can't determine whether Maple tap/inputformats content to the old or new API. Work around that.

Commit:
PlatformTest.IterableSource test: remove incorrect implicit ordering assumption (which is not upheld in Tez)

Commit:
Ensure at least SOME test also run on Tez

Commit:
Reduce the amount of logging noise from Tez during tests

Commit:
Add cascading-hadoop back to the assembly used by REPL

Commit:
tracking Cascading 3.1-wip-48

Commit:
more test output into .gitignore

Commit:
(-commons: post-merge with upstream/develop, lost a "provided" dep on cascading-hadoop)

Commit:
re-C3'ing the VersionedTap/KeyValueByteScheme that were absorbed from dfs-datastores

Commit:
(follow with Cascading, 3.1-wip-52 + tez 0.8.2; and with Scalding/develop)

johnynek commented:
I think #1586 may actually be getting close to merge... I'm closing this in favor of that.

If that is not right. Reopen. Thanks a ton @cchepelov for kicking this all off. I'm sorry it has taken so long.

cchepelov commented:
You're most welcome. I'm relieved and excited this moves forwards.

Will check what's in #1586 and what might be missing and'll forward port 
what I need to switch. Glad to be soon out of the corner I'm painted in ;)


Le 25 janvier 2018 7:52:04 PM "P. Oscar Boykin" <notifications@github.com> 
a écrit :

> I think #1586 may actually be getting close to merge... I'm closing this in 
> favor of that.
>
> If that is not right. Reopen. Thanks a ton @cchepelov for kicking this all 
> off. I'm sorry it has taken so long.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/twitter/scalding/pull/1446#issuecomment-360562721
